Library/Homebrew/cmd/upgrade.rb

```diff
@@ -191,8 +195,11 @@ def upgrade_outdated_formulae(formulae, args:)
       end
     end
 
-    if !pinned.empty? && !args.ignore_pinned?
-      ofail "Not upgrading #{pinned.count} pinned #{Utils.pluralize("package", pinned.count)}:"
+    if pinned.any?
+      Kernel.public_send(
+        formulae.any? ? :ofail : :opoo, # only fail when pinned formulae are named explicitly
+        "Not upgrading #{pinned.count} pinned #{Utils.pluralize("package", pinned.count)}:",
+      )
```

Comment:
I handled the case where we want the command to error if a pinned package is explicitly passed to `brew upgrade` but not when it's called without package arguments (https://github.com/Homebrew/brew/pull/16301/commits/9a00f2795d9eca48385a5222d953787b0269f050).

```console
$ brew upgrade
Warning: Not upgrading 1 pinned package:
fish 3.6.4
==> No packages to upgrade
$ brew upgrade fish
Error: Not upgrading 1 pinned package:
fish 3.6.4
==> No packages to upgrade
```